controller/iteration.go

```diff
@@ -232,6 +233,92 @@ func (c *IterationController) Update(ctx *app.UpdateIterationContext) error {
 	})
 }
 
+// Delete runs the delete action.
+func (c *IterationController) Delete(ctx *app.DeleteIterationContext) error {
+	currentUser, err := login.ContextIdentity(ctx)
+	if err != nil {
+		return jsonapi.JSONErrorResponse(ctx, goa.ErrUnauthorized(err.Error()))
+	}
+	err = application.Transactional(c.db, func(appl application.Application) error {
+		itr, err := appl.Iterations().Load(ctx.Context, ctx.IterationID)
+		if err != nil {
+			return jsonapi.JSONErrorResponse(ctx, err)
+		}
+		s, err := appl.Spaces().Load(ctx, itr.SpaceID)
+		if err != nil {
+			return jsonapi.JSONErrorResponse(ctx, goa.ErrNotFound(err.Error()))
+		}
+		if !uuid.Equal(*currentUser, s.OwnerId) {
+			log.Warn(ctx, map[string]interface{}{
+				"space_id":     s.ID,
+				"space_owner":  s.OwnerId,
+				"current_user": *currentUser,
+			}, "user is not the space owner")
+			return jsonapi.JSONErrorResponse(ctx, errors.NewForbiddenError("user is not the space owner"))
+		}
+		if c.IsRoot(itr, s) {
+			log.Warn(ctx, map[string]interface{}{
+				"space_id":     s.ID,
+				"iteration_id": itr.ID,
+			}, "can not delete root iteration")
+			return jsonapi.JSONErrorResponse(ctx, errors.NewForbiddenError("can not delete root iteration"))
+		}
+		subtree, err := appl.Iterations().LoadChildren(ctx, ctx.IterationID)
+		if err != nil {
+			return jsonapi.JSONErrorResponse(ctx, goa.ErrInternal(err.Error()))
+		}
+		// delete all children along with given iteration
+		subtree = append(subtree, *itr)
```

Comment:
@pranavgore09 Shouldn't we first associate the workitems from all children to the root and THEN delete the iteration?

Reply:
thanks, @DhritiShikhar, yeah I thought that while writing, but they are in the same transaction and it is not foreign key so I am keeping the order as it is.

Reply:
@baijum what do you think in such case should we worry about the order of queries being fired?

Reply:
I would follow @DhritiShikhar because Postgres might follow the chronological order of SQL statements. Also, from a reviewer's point of view, it's easier to understand the code if you move the work items first and then delete the sub-iterations.

Reply:
also, if there were some cascade deletes at the DB level, all work items might be deleted before being moved (it is not the case for now)

Reply:
thanks @DhritiShikhar for bringing this up and @xcoulon for clarification, now updated the order in  
https://github.com/fabric8-services/fabric8-wit/pull/1683/commits/21a922cabd1ddca24311e954ec69455efd52b824